### PR TITLE
Implement roles

### PR DESCRIPTION
**Pull request by akhenry**
Closes https://github.com/akhenry/openmct-yamcs/issues/132
Closes https://github.com/akhenry/openmct-yamcs/issues/130

To Do
* [x] ~~Document how to configure XTCE and roles in Yamcs to support Operator Status in README.md~~. - Have filed followup - https://github.com/akhenry/openmct-yamcs/issues/135

### Describe your changes:
* Implements the user status provider to allow Yamcs to be a source of Operator Status information.
* Converts the Open MCT-Yamcs project to Webpack 5 to support private member variables
* Cleans up deprecated Yamcs configuration
* Updates the example project to use Operator Status
* Implements full support for Yamcs enums
* Implements a very rudimentary "latest data provider" that uses the [`get parameter value` API call](https://docs.yamcs.org/yamcs-http-api/processing/get-parameter-value/) to fetch the latest known operator status value from the real-time processor. This provider will likely be extended in future [to batch fetch values for all latest telemetry requests](https://github.com/akhenry/openmct-yamcs/issues/124).
* Extends the Yamcs user provider to also return user roles from Yamcs. 

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [ ] Unit tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [ ] Testing instructions included in associated issue?

### Reviewer Checklist

* [x] Changes appear to address issue?
* [x] Changes appear not to be breaking changes?
* [ ] Appropriate unit tests included?
* [x] Code style and in-line documentation are appropriate?
* [x] Commit messages meet standards?
* [x] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [x] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

### CODE REVIEW

File: src/providers/user/operator-status-telemetry.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    async getPossibleStatuses() {
+        return this.#readyPromise.then(() => {
+            return Object.values(this.#statusMap).map(status => this.toStatusFromMdbEntry(status));
+        });
+    }
+    addStatus(status) {
+        this.#statusMap[status.value] = status;
+    }
+    setPossibleStatusesForRole(role, possibleStates) {
+        this.#statusMap[role] = possibleStates;
```
Comment (shefalijoshi) [inline src/providers/user/operator-status-telemetry.js:81]: The `addStatus` method above this one is assigning a single `status` for each `#statusMap` entry, but here we're setting an array of `possibleStates` for each `#statusMap` entry. That's confusing to me - is it just a variable name issue?

**Review by shefalijoshi — CHANGES_REQUESTED** [review]
That's a lot of work! I think I understand most of it - did the first pass, but will come back to it again.

File: example/index.js [hunk 2]
```diff
@@ -4,11 +4,38 @@ import installYamcsPlugin from '../src/plugin.js';
 const config = {
     "yamcsDictionaryEndpoint": "http://localhost:9000/yamcs-proxy/",
     "yamcsHistoricalEndpoint": "http://localhost:9000/yamcs-proxy/",
-    "yamcsRealtimeEndpoint": "ws://localhost:9000/yamcs-proxy-ws/",
-    "yamcsWebsocketEndpoint": "ws://localhost:9000/yamcs-proxy-ws-v2/",
+    "yamcsWebsocketEndpoint": "ws://localhost:9000/yamcs-proxy-ws/",
+    "yamcsUserEndpoint": "http://localhost:9000/yamcs-proxy/api/user",
```
[the remark this answers is not in the document]
Comment (akhenry) [reply example/index.js]: Fixed

File: package.json [hunk 3]
```diff
@@ -8,7 +8,6 @@
     "lint": "eslint src example",
     "lint:fix": "eslint src example --fix",
     "build:dist": "NODE_ENV=production webpack",
-    "prepublish": "npm install openmct@unstable --no-save",
```
[the remark this answers is not in the document]
Comment (akhenry) [reply package.json:11]: Fixed

File: src/providers/historical-telemetry-provider.js [hunk 4]
```diff
@@ -113,7 +121,6 @@ export default class YamcsHistoricalTelemetryProvider {
                 && domainObject.type !== OBJECT_TYPES.AGGREGATE_TELEMETRY_TYPE
             ) {
                 options.sizeType = 'count';
-                options.isSamples = true;
```
[the remark this answers is not in the document]
Comment (akhenry) [reply src/providers/historical-telemetry-provider.js:116]: `false` is the default for most cases. It gets overridden in `request()` though for minmax requests. It was being triggered in a number of places it shouldn't have been.

File: src/providers/latest-telemetry-provider.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        const url = this._buildUrl(domainObject.identifier);
+        const response = await fetch(url);
+
+        let result = await response.json();
+        let openMctStyleDatum = undefined;
+
+        if (result !== undefined) {
+            if (result.acquisitionStatus !== undefined) {
+                openMctStyleDatum = {
+                    id: result.id.name,
```
[the remark this answers is not in the document]
Comment (akhenry) [reply src/providers/latest-telemetry-provider.js:43]: If it doesn't exist then Yamcs has returned invalid data and it's an error.

File: src/providers/user/createYamcsUser.js [hunk 6]
```diff
@@ -37,6 +39,14 @@ export default function createYamcsUser(UserClass) {
             this.creationTime = creationTime;
             this.confirmationTime = confirmationTime;
             this.lastLoginTime = lastLoginTime;
+            this.roles = roles.map(role => role.name);
+            this.objectPrivileges = objectPrivilege;
```
[the remark this answers is not in the document]
Comment (akhenry) [reply src/providers/user/createYamcsUser.js:43]: It's not great is it. [This is how it comes back from Yamcs unfortunately](https://github.com/akhenry/openmct-yamcs/blob/4a9597d777a0523e1c5f38aa4b025acccb193215/src/providers/user/user-provider.js#L165). I'll add some jsdocs to make it clearer.

File: src/providers/user/operator-status-parameter.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        const aliases = parameter.alias;
+
+        return aliases !== undefined
+            && aliases.some(alias => alias.name === OPERATOR_STATUS_TYPE);
+    }
+
+    getRoleFromParameter(parameter) {
+        const aliases = parameter.alias;
+
+        return aliases.find(alias => alias.namespace === 'OpenMCT:role').name;
```
[the remark this answers is not in the document]
Comment (akhenry) [reply src/providers/user/operator-status-parameter.js:36]: In the calling code this is only ever called on parameters that have a role alias in the XTCE definition (the previous call to `isOperatorStatusParameter` checks this). If so, it must also define a role, so it's an error if `.find()` returns nothing. I will make this error explicit to help developers.

File: src/providers/user/operator-status-telemetry.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        this.#statusRoles = new Set();
+        this.#roleToTelemetryObjectMap = {};
+        this.#readyPromise = new Promise((resolve) => this.#setReady = resolve);
+        this.#url = url;
+        this.#instance = instance;
+        this.#processor = processor;
+        this.#openmct = openmct;
+    }
+    async setStatusForRole(role, status) {
+        //TODO Error handling.
```
[the remark this answers is not in the document]
Comment (akhenry) [reply src/providers/user/operator-status-telemetry.js:48]: Error handling is implemented, good catch.

File: src/providers/user/operator-status-telemetry.js (hunk 1 above)
Comment (akhenry) [reply src/providers/user/operator-status-telemetry.js:81]: Good catch, thanks. This is actually code from a previous iteration that is no longer used.

**Comment by akhenry** [comment]
@shefalijoshi OK, I believe this is ready to go again. Thanks for the review!

**Review by shefalijoshi — APPROVED** [review]
Lovely!

**Pull request merged**